gson/src/test/java/com/google/gson/functional/DefaultTypeAdaptersTest.java

```diff
@@ -812,6 +836,232 @@ public void testStringBufferDeserialization() {
+            + "\"}}";
+    roundTrip(zonedDateTime, json);
+  }
+
+  @Test
+  public void testJavaTimeZonedDateTimeWithZoneIdThatHasAdapter() {
+    TypeAdapter<ZoneId> zoneIdAdapter =
+        new TypeAdapter<ZoneId>() {
+          @Override
+          public void write(JsonWriter out, ZoneId value) throws IOException {
+            out.value(value.getId());
+          }
+
+          @Override
+          public ZoneId read(JsonReader in) throws IOException {
+            return ZoneId.of(in.nextString());
+          }
+        };
+    Gson customGson = new GsonBuilder().registerTypeAdapter(ZoneId.class, zoneIdAdapter).create();
+    ZoneId zoneId = ZoneId.of("UTC+01:00");
+    int totalSeconds = ((ZoneOffset) zoneId.normalized()).getTotalSeconds();
+    ZonedDateTime zonedDateTime =
+        ZonedDateTime.of(LocalDate.of(2021, 12, 2), LocalTime.of(12, 34, 56, 789_012_345), zoneId);
+    String json =
+        "{\"dateTime\":{\"date\":{\"year\":2021,\"month\":12,\"day\":2},"
+            + "\"time\":{\"hour\":12,\"minute\":34,\"second\":56,\"nano\":789012345}},"
+            + "\"offset\":{\"totalSeconds\":"
+            + totalSeconds
+            + "},"
+            + "\"zone\":\""
+            + zoneId.getId()
+            + "\"}";
+    roundTrip(customGson, zonedDateTime, ZonedDateTime.class, json);
+  }
+
+  private static final boolean JAVA_TIME_FIELDS_ARE_ACCESSIBLE;
+
+  static {
+    boolean accessible = false;
+    try {
+      Instant.class.getDeclaredField("seconds").setAccessible(true);
+      accessible = true;
+    } catch (InaccessibleObjectException e) {
+      // OK: we can't reflect on java.time fields
+    } catch (NoSuchFieldException e) {
+      // JDK implementation has changed and we no longer have an Instant.seconds field.
+      throw new AssertionError(e);
+    }
+    JAVA_TIME_FIELDS_ARE_ACCESSIBLE = accessible;
+  }
+
+  private void roundTrip(Object value, String expectedJson) {
+    roundTrip(value, value.getClass(), expectedJson);
+  }
+
+  private void roundTrip(Object value, Class<?> valueClass, String expectedJson) {
+    roundTrip(gson, value, valueClass, expectedJson);
+    if (JAVA_TIME_FIELDS_ARE_ACCESSIBLE) {
+      checkReflectiveTypeAdapterFactory(value, expectedJson);
+    }
+  }
+
+  private void roundTrip(Gson customGson, Object value, Class<?> valueClass, String expectedJson) {
+    assertThat(customGson.getAdapter(valueClass).getClass().getName()).doesNotContain("Reflective");
+    assertThat(customGson.toJson(value, valueClass)).isEqualTo(expectedJson);
+    assertThat(customGson.fromJson(expectedJson, valueClass)).isEqualTo(value);
+  }
+
+  // Assuming we have reflective access to the fields of java.time classes, check that
+  // ReflectiveTypeAdapterFactory would produce the same JSON. This ensures that we are preserving
+  // a compatible JSON format for those classes even though we no longer use reflection.
+  private void checkReflectiveTypeAdapterFactory(Object value, String expectedJson) {
```

Comment:
Do you want to skip the tests yet under Java 26(?)+, or is it too early to worry about that?

Reply:
I think I'll address that when it comes up. There's already a lot going on in this PR. :-)